sdc/native/hashmap.hpp

```diff
@@ -0,0 +1,907 @@
+
+#include <cstdint>
+#include <iostream>
+#include <utility>
+#include <memory>
+#include <type_traits>
+
+#include "utils.hpp"
+#include "tbb/tbb.h"
+#include "tbb/concurrent_unordered_map.h"
+#include "numba/core/runtime/nrt_external.h"
+
+
+using voidptr_hash_type = size_t (*)(void* key_ptr);
+using voidptr_eq_type = bool (*)(void* lhs_ptr, void* rhs_ptr);
+using voidptr_refcnt = void (*)(void* key_ptr);
+
+using iter_state = std::pair<void*, void*>;
+
+
+class CustomVoidPtrHasher
+{
+private:
+    voidptr_hash_type ptr_hash_callback;
+
+public:
+    CustomVoidPtrHasher(void* ptr_func) {
+        ptr_hash_callback = reinterpret_cast<voidptr_hash_type>(ptr_func);
+    }
+
+    size_t operator()(void* data_ptr) const {
+        auto res = ptr_hash_callback(data_ptr);
+        return res;
+    }
+};
+
+
+class CustomVoidPtrEquality
+{
+private:
+    voidptr_eq_type ptr_eq_callback;
+
+public:
+    CustomVoidPtrEquality(void* ptr_func) {
+        ptr_eq_callback = reinterpret_cast<voidptr_eq_type>(ptr_func);
+    }
+
+    size_t operator()(void* lhs, void* rhs) const {
+        return ptr_eq_callback(lhs, rhs);
+    }
+};
+
+
+struct VoidPtrHashCompare {
+private:
+    voidptr_hash_type ptr_hash_callback;
+    voidptr_eq_type ptr_eq_callback;
+
+public:
+    size_t hash(void* data_ptr) const {
+        return ptr_hash_callback(data_ptr);
+    }
+    bool equal(void* lhs, void* rhs) const {
+        return ptr_eq_callback(lhs, rhs);
+    }
+
+    VoidPtrHashCompare(void* ptr_hash, void* ptr_equality) {
+        ptr_hash_callback = reinterpret_cast<voidptr_hash_type>(ptr_hash);
+        ptr_eq_callback = reinterpret_cast<voidptr_eq_type>(ptr_equality);
+    }
+
+    VoidPtrHashCompare() = delete;
+    VoidPtrHashCompare(const VoidPtrHashCompare& rhs) = default;
+    VoidPtrHashCompare& operator=(const VoidPtrHashCompare&) = default;
+    VoidPtrHashCompare& operator=(VoidPtrHashCompare&&) = default;
+    ~VoidPtrHashCompare() = default;
+};
+
+
+struct VoidPtrTypeInfo {
+    voidptr_refcnt incref;
+    voidptr_refcnt decref;
+    uint64_t size;
+
+    VoidPtrTypeInfo(void* incref_addr, void* decref_addr, uint64_t val_size) {
+        incref = reinterpret_cast<voidptr_refcnt>(incref_addr);
+        decref = reinterpret_cast<voidptr_refcnt>(decref_addr);
+        size = val_size;
+    }
+
+    VoidPtrTypeInfo() = delete;
+    VoidPtrTypeInfo(const VoidPtrTypeInfo&) = default;
+    VoidPtrTypeInfo& operator=(const VoidPtrTypeInfo&) = default;
+    VoidPtrTypeInfo& operator=(VoidPtrTypeInfo&&) = default;
+    ~VoidPtrTypeInfo() = default;
+};
+
+
+template<typename Key,
+         typename Val,
+         typename Hasher=std::hash<Key>,
+         typename Equality=std::equal_to<Key>
+>
+class NumericHashmapType {
+public:
+    using map_type = typename tbb::concurrent_unordered_map<Key, Val, Hasher, Equality>;
+    using iterator_type = typename map_type::iterator;
+    std::unique_ptr<map_type> p_map;
```

Comment:
Why to store it as pointer, not as value?

Reply:
Done.